src/Dfc.CourseDirectory.Core/BinaryStorageProvider/BlobStorageBinaryStorageProvider.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System.IO;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Azure.Storage.Blobs;
+using Azure;
+
+namespace Dfc.CourseDirectory.Core.BinaryStorageProvider
+{
+    public class BlobStorageBinaryStorageProvider : IBinaryStorageProvider
+    {
+        private readonly BlobContainerClient _blobContainerClient;
+        private bool _containerExists;
+
+        public BlobStorageBinaryStorageProvider(IOptions<BlobStorageBinaryStorageProviderSettings> options)
+        {
+            _blobContainerClient = new BlobContainerClient(options.Value.ConnectionString, options.Value.ContainerName);
+        }
+
+        public async Task<bool> TryDownloadFile(string path, Stream destination)
+        {
+            await EnsureContainerExists();
+
+            var blob = _blobContainerClient.GetBlobClient(path);
+
+            try
+            {
+                await blob.DownloadToAsync(destination);
```

Comment:
I'm not sure of your intended consumption of this method, however it might be more flexible if you were to return the clients stream directly rather than copying onto a new stream...

```
public async Task<(bool success, Stream result)> Download(string blobName)
{
    var blob = _blobContainerClient.GetBlobClient(path);

    try
    {
        return (true, (await blobClient.DownloadAsync()).Value.Content);
    }
    catch (RequestFailedException ex)
        when (ex.Status == StatusCodes.Status404NotFound)
    {
        return (false, null);
    }
}
```

Reply:
My thinking behind current signature was you could stream a blob directly to an `HttpResponse`'s body without ever having the whole thing in memory.

Reply:
If that's your use case then I'd certainly suggest using `DownloadAync()` and grabbing `Value.Content`. With your original approach you're going to need to new-up a `MemoryStream` to pass as a parameter to the method which will involve writing it to memory.

Reply:
NB the value tuple in my suggestion isn't a great idea! :)

Reply:
Why would you need a `MemoryStream` at all? You can write directly to `HttpResponse.Body` (or do it via an `IActionResult`)

Reply:
When calling `TryDownloadFile(string path, Stream destination)` you need a pass an initialised Stream object which has to come from somewhere. (Please correct me if I'm wrong, but I don't believe `HttpResponseMessage `or any of the `IActionResult `implementations provide you with an initialised `Stream` object to pass.)